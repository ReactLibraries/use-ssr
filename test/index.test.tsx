import React, { useEffect, useState } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { reset } from '@react-libraries/use-global-state';
import { createCache, getDataFromTree, useSSR } from '../src';

let container: HTMLElement;
beforeEach(() => {
  reset();
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
});

const Component01 = () => {
  const [count, setCount] = useState(0);
  const [state, setState] = useSSR<number>('Component01', async (state, setState) => {
    if (state === 100) return;
    setState(10);
    setCount((v) => v + 1);
    setState(100);
  });
  useEffect(() => {
    setState(200);
  }, []);
  return <>{[state, count]}</>;
};
const Component02 = () => {
  const [count, setCount] = useState(0);
  const [state, setState] = useSSR<number>(['Component', '02'], async (state, setState) => {
    if (state !== undefined) return;
    setState(10);
    await new Promise((resolve) => setTimeout(resolve, 1));
    setState(100);
    setCount((v) => v + 1);
  });
  useEffect(() => {
    setState(undefined);
  }, []);
  return <>{[state, count]}</>;
};
const Component03 = () => {
  const [state] = useSSR<number>(['Component', '02']);
  return <>{[state]}</>;
};

it('Client', async () => {
  await act(async () => {
    const cache = await getDataFromTree(<Component01 />);
    expect(cache).toMatchSnapshot();
    render(
      <>
        <Component01 />
      </>,
      container
    );
  });
  expect(container.childNodes).toMatchSnapshot();

  unmountComponentAtNode(container);
  container.remove();

  await act(async () => {
    const cache = await getDataFromTree(<Component01 />);
    createCache(cache);
    render(
      <>
        <Component01 />
      </>,
      container
    );
  });
  expect(container.childNodes).toMatchSnapshot();

  unmountComponentAtNode(container);
  container.remove();

  await act(async () => {
    const cache = await getDataFromTree(<></>);
    expect(cache).toMatchSnapshot();
    createCache(cache);
    render(
      <>
        <Component01 />
        <Component02 />
      </>,
      container
    );
    await new Promise((r) => setTimeout(r, 100));
  });
  expect(container.childNodes).toMatchSnapshot();
  unmountComponentAtNode(container);
  container.remove();

  await act(async () => {
    const cache = await getDataFromTree(
      <>
        <Component01 />
        <Component02 />
      </>
    );
    expect(cache).toMatchSnapshot();
    createCache(cache);
    render(
      <>
        <Component01 />
        <Component02 />
      </>,
      container
    );
    await new Promise((r) => setTimeout(r, 100));
  });

  expect(container.childNodes).toMatchSnapshot();
});

it('SSR', async () => {
  (process as { browser?: boolean }).browser = true;
  await act(async () => {
    const cache = await getDataFromTree(
      <>
        <Component01 />
        <Component02 />
        <Component03 />
      </>
    );
    expect(cache).toMatchSnapshot();
    createCache(cache);
    render(
      <>
        <Component01 />
        <Component02 />
        <Component03 />
      </>,
      container
    );
    await new Promise((r) => setTimeout(r, 100));
  });

  expect(container.childNodes).toMatchSnapshot();
});
