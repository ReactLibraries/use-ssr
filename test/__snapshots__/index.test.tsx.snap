// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clear 1`] = `
<div>
  1000
</div>
`;

exports[`Clear2 1`] = `<div />`;

exports[`Client 1`] = `
Object {
  "[@react-libraries/use-ssr][Component01]": 100,
}
`;

exports[`Client 2`] = `
NodeList [
  100,2,
]
`;

exports[`Client 3`] = `
NodeList [
  100,1,
]
`;

exports[`Client 4`] = `Object {}`;

exports[`Client 5`] = `
NodeList [
  100,1,
  100,1,
]
`;

exports[`Client 6`] = `
Object {
  "[@react-libraries/use-ssr][Component01]": 100,
  "[@react-libraries/use-ssr][Component][02]": 100,
}
`;

exports[`Client 7`] = `
NodeList [
  100,1,
  100,1,
]
`;

exports[`Context 1`] = `
Object {
  "[@react-libraries/use-ssr][Component01]": 100,
  "[@react-libraries/use-ssr][Component][02]": 1000,
}
`;

exports[`Context 2`] = `
NodeList [
  100,1,
  100,1,
  100,
]
`;

exports[`Mutation-Query 1`] = `Object {}`;

exports[`Mutation-Query 2`] = `
NodeList [
  100,3,
  123,1,
  123,
  100,,
]
`;

exports[`SSR 1`] = `
Object {
  "[@react-libraries/use-ssr][Component01]": 100,
  "[@react-libraries/use-ssr][Component][02]": 100,
}
`;

exports[`SSR 2`] = `
NodeList [
  100,1,
  100,1,
  100,
]
`;
